### PR TITLE
#164139688: Comment on articles 

### PR DESCRIPTION
**Pull request by olajide1234**
#### What does this PR do?
- Includes ability to comment on articles

#### Description of Task to be completed?
- Create `comment` route
- Create `post`, `edit` and `delete` methods on the route

#### How should this be manually tested?
- Run `npm run start` to start the application.
- Make a `POST` request to `/api/v1/articles/:slug/comments`, include slug of article that comment will be associated with as params; include `comment` field in body of request.
- Make an `GET` request to `/api/v1/articles/:slug/comments/:id`,  include slug of article to fetch comment for as params
- Make an `EDIT` request to `/api/v1/articles/:slug/comments/:id`,  include id of comment to be edited as params;  include `comment` field in body of request.
- Make a `DELETE` request to `/api/v1/articles/:slug/comments/:id`,  include id of comment to be deleted as params.

### What are the relevant pivotal tracker stories?
- [#164139688](https://www.pivotaltracker.com/story/show/164139688)

#### Any background context you want to provide?
- None

#### Screenshots (if appropriate)
- None

#### Questions:
- N/A

### CODE REVIEW

File: routes/v1.js [hunk 1]
```diff
@@ -92,5 +99,24 @@ apiRoutes.post(
   returnValidationErrors,
   UserController.resetPassword
 );
+apiRoutes.route('/comment/:id')
```
Comment (kleva-j) [inline routes/v1.js]: Yusuf, I think we should follow the docs on the README.MD file. The route to making a comment should be like this `POST /api/articles/:slug/comments`

File: controllers/comment.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      });
+      return res.status(205).json({
+        success: true,
+        message: 'Comments returned successfully',
+        comments: findComment,
+      });
+    } catch (error) {
+      return res.status(500).json({
+        success: false,
+        errors: [error.message]
```
Comment (chingsley) [inline controllers/comment.js:76]: confirm that the error object has a message property.

**Review by kleva-j — APPROVED** [review]

**Review by chingsley — APPROVED** [review]

File: routes/v1.js (hunk 1 above)
Comment (olajide1234) [reply routes/v1.js]: Updated. Thanks.

File: controllers/comment.js (hunk 2 above)
Comment (olajide1234) [reply controllers/comment.js:76]: That seems to be the standard error format. See more here: https://github.com/sequelize/sequelize/blob/3e5b8772ef75169685fc96024366bca9958fee63/lib/errors.js#L49

File: controllers/comment.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+   * @param {object} req - HTTP Request
+   * @param {object} res - HTTP Response
+   *
+   * @memberof CommentController
+   *
+   * @returns {object} Article comment
+   */
+  static async createComment(req, res) {
+    const { id } = req.user;
+    const { slug: articleSlug } = req.params;
```
Comment (9jaswag) [inline controllers/comment.js:23]: @olajide1234 I'm not sure the param used by others, please confirm and ensure it's uniform.

File: routes/v1.js [hunk 4]
```diff
@@ -104,4 +109,38 @@ apiRoutes.post(
...
+  )
+  .get(
+    Auth.verifyUser,
+    isUserVerified,
+    validateArticleExist,
+    CommentController.getComments
+  );
+
+apiRoutes.route('/articles/:slug/comments/:id')
+  .patch(
```
Comment (Dubby20) [inline routes/v1.js:181]: Please change your http method to `put`

File: test/comment.spec.js [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        done(err);
+      });
+  });
+});
+describe('Edit a comment', () => {
+  before(() => updateVerifiedStatus(validCommentUser.email));
+  it('returns successfully edit comment', (done) => {
+    chai
+      .request(app)
+      .patch(`/api/v1/articles/:slug/comments/${commentId}`)
```
Comment (Dubby20) [inline test/comment.spec.js:196]: Change to `put`

File: routes/v1.js (hunk 4 above)
Comment (olajide1234) [reply routes/v1.js:181]: As discussed, `patch` seems to be the more appropriate method for this. Thanks.

File: test/comment.spec.js (hunk 5 above)
Comment (olajide1234) [reply test/comment.spec.js:196]: As discussed, `patch` seems to be the more appropriate method for this. Thanks.

File: controllers/comment.js (hunk 3 above)
Comment (olajide1234) [reply controllers/comment.js:23]: I have confirmed, the team uses `req.params.slug` to pass the slug of articles. I have only reassigned the name to `articleSlug` in my controller.

File: middleware/validation.js [hunk 6]
```diff
@@ -141,3 +141,80 @@ export const validatePassword = [
...
     .withMessage('No spaces are allowed in the password.')
 ];
+export const validateCreateComment = [
+  check('comment')
+    .exists()
+    .withMessage('Please include a comment in the body of request.')
+    .isString()
+    .trim()
+    .withMessage('Comment must be alphanumeric characters, please remove leading and trailing whitespaces.')
+    .isLength({ min: 2 })
```
Comment (9jaswag) [inline middleware/validation.js:157]: @olajide1234 did you guys agree on a min length for this?

Comment (olajide1234) [reply middleware/validation.js:157]: Yes, we have agreed min length for comments should be 2 characters.

**Comment by 9jaswag** [comment]
@olajide1234 you have Michael's commit in yours. Plus some merge conflicts.

**Comment by olajide1234** [comment]
> @olajide1234 you have Michael's commit in yours. Plus some merge conflicts.

Yes, i will remove them now and rebase to fix the conflict.

**Comment by 9jaswag** [comment]
@olajide1234 please only re-add me when all the required tests have passed.

**Comment by olajide1234** [comment]
> @olajide1234 please only re-add me when all the required tests have passed.

All test are now passing.

**Pull request merged**